connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java

```diff
@@ -343,12 +336,38 @@ public Future<Void> set(final Map<ByteBuffer, ByteBuffer> values, final Callback
         return producerCallback;
     }
 
+    @Override
+    public Set<Map<String, Object>> connectorPartitions(String connectorName) {
+        return connectorPartitions.getOrDefault(connectorName, Collections.emptySet());
+    }
+
+    @SuppressWarnings("unchecked")
     protected final Callback<ConsumerRecord<byte[], byte[]>> consumedCallback = (error, record) -> {
         if (error != null) {
             log.error("Failed to read from the offsets topic", error);
             return;
         }
 
+        if (record.key() != null) {
+            try {
+                // The key should always be a list of the form [connectorName, partition] where connectorName is a
+                // string value and partition is a Map<String, Object>
+                List<Object> keyValue = (List<Object>) keyConverter.toConnectData(topic, record.key()).value();
+                String connectorName = (String) keyValue.get(0);
+                Map<String, Object> partition = (Map<String, Object>) keyValue.get(1);
```

Comment:
I know it's a bit of a pain, but we should be more careful about casting here. If the deserialized value isn't a list, we should log a warning and move on; if it doesn't have two elements, log and move on; basically, no data with an unexpected format should break anything here.

This is mostly to leave room for evolution of the format of records in the offsets topic without worrying about breaking workers running older versions during rolling upgrades or emergency downgrades.

Reply:
Thanks, that's a good example. I've refactored this to have more fine grained error messages and moved it to `OffsetUtils` for re-use (also added a few simple tests while I was at it).